src/client.ts

```diff
@@ -9297,31 +9297,60 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
 
     /**
      * Find the event_id closest to the given timestamp in the given direction.
-     * @return {Promise} A promise of an object containing the event_id and
-     *    origin_server_ts of the closest event to the timestamp in the given
-     *    direction
+     * @return {Promise} Resolves: A promise of an object containing the event_id and
+     *    origin_server_ts of the closest event to the timestamp in the given direction
+     * @return {module:http-api.MatrixError} Rejects: when the request fails
      */
-    public timestampToEvent(
+    public async timestampToEvent(
         roomId: string,
         timestamp: number,
         dir: Direction,
     ): Promise<ITimestampToEventResponse> {
         const path = utils.encodeUri("/rooms/$roomId/timestamp_to_event", {
             $roomId: roomId,
         });
+        const queryParams = {
+            ts: timestamp.toString(),
+            dir: dir,
+        };
 
-        return this.http.authedRequest(
-            Method.Get,
-            path,
-            {
-                ts: timestamp.toString(),
-                dir: dir,
-            },
-            undefined,
-            {
-                prefix: "/_matrix/client/unstable/org.matrix.msc3030",
-            },
-        );
+        try {
+            return await this.http.authedRequest(
+                Method.Get,
+                path,
+                queryParams,
+                undefined,
+                {
+                    prefix: ClientPrefix.V1,
+                },
+            );
+        } catch (err) {
+            // Fallback to the prefixed unstable endpoint. Since the stable endpoint is
+            // new, we should also try the unstable endpoint before giving up. We can
+            // remove this fallback request in a year (remove after 2023-11-28).
+            if (
+                (<MatrixError>err).errcode === "M_UNRECOGNIZED" && (
+                    // XXX: The 400 status code check should be removed in the future
+                    // when Synapse is compliant with MSC3743.
+                    (<MatrixError>err).httpStatus === 400 ||
```

Comment:
I'd somewhat suggest not going for lowest common denominator here and just look for MSC3743 instead. Synapse should have support as of recently, and the spec says MSC3743 is stable, and for the feature this impacts I'd expect the user to upgrade both their Element and Synapse before trying to use it (or their admin will do the update, at least).

Reply:
All versions of Synapse that had unstable [MSC3030](https://github.com/matrix-org/matrix-spec-proposals/pull/3030) support, didn't support [MSC3743](https://github.com/matrix-org/matrix-spec-proposals/pull/3743). If we removed falling back when `400`, that means that the fallback would never be called and we might as well remove the fallback altogether.

Reply:
That sounds regrettable, but to be honest it's pretty appealing to not have a fallback then. We're primarily a Matrix client, not a Synapse client, and 400 is a really awful error code to do logic like this on, so instead there's likely a path where we just don't fall back and point at MSC3743 for not existing sooner.

I'm somewhat easily convinced on either side though, if you have thoughts on the matter.

Reply:
Opting to leave the fallback in place for now ⏩